zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java

```diff
@@ -51,7 +57,18 @@
 public class X509AuthenticationProvider implements AuthenticationProvider {
```

Comment:
We're doing quite a few customizations to `X509AuthenticationProvider` and can lead to merge conflicts in the future. Would it make more sense to create a new AuthenticationProvider (or maybe inherit from `X509AuthenticationProvider`) with all these customizations?

Reply:
This isn't that much customization considering both logic paths are based on X509 Certificates. This way, we avoid boilerplate. We will consider a different implementation if we consider it necessary (for example, another auth mechanism that integrates with a 3rd party library/service or something that uses more than just X509 certs).